docs/cip/CIP-07102024_Write_Ahead_Log_Pruning_Vacuuming.md

```diff
@@ -0,0 +1,64 @@
+# CIP-07102024: Write-Ahead Log Pruning & Vacuuming
+
+## Status
+
+Current Status: `Under Discussion`
+
+## Motivation
+
+Chroma's SQLite-based write-ahead log grows infinitely over time. When ingesting large amounts of data, it's not uncommon for the SQLite database to grow to many gigabytes in size. Large databases cost more, take longer to back up, and can result in decreased query performance.
+
+There are two separate problems:
+
+- The database, specifically the `embeddings_queue` table, has unbounded growth.
+- The SQLite `VACUUM` command, often recommended for such scenarios, is a blocking and potentially slow operation. Read and write operations are both blocked during a `VACUUM`.[^1]
+
+This CIP addresses both issues.
+
+## Proposed Changes
+
+A new configuration parameter will be added, `log:vacuum_threshold`. It defaults to 1GB. Following Postgres' convention, the unit is megabytes. This helps avoid excessive fragmentation—without this parameter, or if it's set to `0`, it is effectively the same as SQLite's full vacuum mode.
+
+Two additional things will be done after write transactions:
+
+1. The `embeddings_queue` table will be pruned to remove rows that are no longer needed. Specifically, rows with a sequence ID less than the minimum sequence ID of any active subscriber will be deleted.
```

Comment:
WAL is extremely valuable, as more than once, users have corrupted DBs or accidentally deleted their binary index dirs. Let's thread lightly here and make this an opt-in. Furthermore, let's make users aware of the ramifications of WAL pruning and potentially offer ways or processes to back up the WAL, which can help recovery.

Reply:
I'm not saying to use the WAL as a backup. All I'm saying is that the WAL is currently the only recovery option in Chroma, and users (some) have come to expect that Chroma will automatically rebuild binary indices if they happen to delete them. Enabling auto-pruning by default circumvents an existing behavior and user (again, some) user expectations.

Reply:
This is a valid point on some levels. I view this an instance of https://www.hyrumslaw.com/.

> and users (some) have come to expect that Chroma will automatically rebuild binary indices if they happen to delete them.

While this may be true, and people may be in the habit of deleting database files and expecting them to just get rebuilt from the WAL - I would argue this is absolutely pathological behavior that we should not encourage or support. Manually deleting files and expecting the database to recreate them seems ripe for disaster. If the flow we want to support is rebuilding indices, we should productize that and promote it to the top level. 

> Enabling auto-pruning by default circumvents an existing behavior
Yes there is some undocumented fact that the WAL is infinite, but this should not be relied upon and preserved. I think we shouldn't prevent ourselves from doing the right thing - pruning the wal and preventing unbounded wal growth - in order to preserve users doing the wrong thing - relying on the WAL for backup/restore behavior as opposed to crash recovery.

Reply:
maybe as part of 0.6.0, we can document recommended backup/recovery flows (probably shut down chroma server, make copy of directory unless we have bandwidth to add a `chroma backup` command?)

Reply:
@tazarov please see updated proposal that addresses this